neo/Consensus/ConsensusService.cs

```diff
@@ -231,22 +322,168 @@ private void OnPrepareRequestReceived(ConsensusPayload payload, PrepareRequest m
+                context.State |= ConsensusState.CommitSent;
+
+                system.LocalNode.Tell(new LocalNode.SendDirectly { Inventory = context.MakeCommitAgreement(context.FinalSignatures[context.MyIndex]) });
+                Log($"Commit sent: height={context.BlockIndex} hash={block.Hash} state={context.State}");
+                CheckFinalSignatures();
             }
         }
 
+
+        private void OnCommitAgreement(ConsensusPayload payload, CommitAgreement message)
+        {
+            if (context.FinalSignatures[payload.ValidatorIndex] != null) return;
+
+            Log($"{nameof(OnCommitAgreement)}: height={payload.BlockIndex} hash={context.MakeHeader().Hash.ToString()} view={message.ViewNumber} index={payload.ValidatorIndex}");
+
+            if (!Crypto.Default.VerifySignature(context.MakeHeader().GetHashData(), message.FinalSignature, context.Validators[payload.ValidatorIndex].EncodePoint(false)))
+            {
+                Log($"{nameof(OnCommitAgreement)}: SIGNATURE verification with problem");
+                return;
+            }
+            context.FinalSignatures[payload.ValidatorIndex] = message.FinalSignature;
+            CheckFinalSignatures();
+        }
+
+        private void OnRegeneration(ConsensusPayload payload, Regeneration message)
+        {
+            if (context.State.HasFlag(ConsensusState.CommitSent)) return;
+
+            Log($"{nameof(OnRegeneration)}: height={payload.BlockIndex} view={message.ViewNumber} numberOfPartialSignatures={message.SignedPayloads.Count(p => p != null)} index={payload.ValidatorIndex}");
+
+            uint nValidSignatures = 0;
+            /// <summary>
+            /// Time for checking if speaker really signed this payload
+            /// </summary>
+            if (!CheckPrimaryPayloadSignature(message.PrepareRequestPayload))
+            {
+                Log($"{nameof(OnRegeneration)}: Regenerating primary payload: {message.PrepareRequestPayload.ValidatorIndex} length:{message.SignedPayloads.Length} with a wrong Primary Payload");
+                context.SignedPayloads[payload.ValidatorIndex] = null;
+                return;
+            }
+            nValidSignatures++;
+
+            /// <summary>
+            /// Time for checking all Backups
+            /// </summary>
+            for (int i = 0; i < message.SignedPayloads.Length; i++)
+                if (message.SignedPayloads[i] != null && i != message.PrepareRequestPayload.ValidatorIndex)
+                    if (!Crypto.Default.VerifySignature(message.PrepareRequestPayload.GetHashData(), message.SignedPayloads[i], context.Validators[i].EncodePoint(false)))
+                    {
+                        Log($"{nameof(OnRegeneration)}: Regenerating {i} payload:{message.PrepareRequestPayload.ValidatorIndex} length:{message.SignedPayloads.Length} is being set to null");
+                        PrintByteArray(message.SignedPayloads[i]);
```

Comment:
We should use the same `Log` method

Reply:
These Logs will be all finished when everything is ready.
When it reaches these points after NGD current tests we come cleaning and adjusting all these.

Reply:
@shargon, do you mean log level?

Reply:
Use the same line for this log

Reply:
I still did not get, ahauauah
Yesterday we cleaned the code, check it out, Shargon.

The Empty PrepareReq Payload was also not fixed yet, what do you recommend?